add more itertools: cartesianPower, combinations, combinationsRepeat

greenify opened the pull request:
As you might have realized I started to enjoy D and come from a Python background (sorry about the contribution "spam"), so I missed some functionality from [Python's itertools](https://docs.python.org/3/library/itertools.html) in Phobos.
This adds a couple of useful combinatoric generators. They are all lazy and provide a length.

### `product`: Cartesian product of `r` with repetitions.

For example `"AB"`.combinationsRepeat(2).array` returns `["AA", "AB", "BA", "BB"]`.

This is different to cartesianProduct in setopts as it allows repeats and avoids to duplicates the input range.

### `combinations`: All k-combinations of `r` with repetitions.

For example `"AB"`.combinations(2).array` returns `["AB"]`.

### `combinationsRepeat`: All k-combinations of `r` with repetitions.

For example `"AB"`.combinationsRepeat(2).array` returns `["AA", "AB", "BB"]`.

There is also [a version](https://github.com/greenify/d-itertools) which combines the three generators in one struct with statics ifs if you prefer this.

greenify commented:
I forgot: I didn't know where to put binomial, actually I was quite surprised it wasn't there. Would `std.math` be a suitable place?

BBasile commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4289,409 @@ unittest
+*/
+auto product(Range)(Range r, size_t repeat=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
```
Phobos brace style is allman

BBasile commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4289,409 @@ unittest
+
+/**
+Lazily computes the all k-combinations of $(D r) with repetitions.
+A k-combination with repetitions, or k-multicombination, or multisubset of size k from a set S is given by a sequence of k not necessarily distinct elements of S, where order is not taken into account.
```
Phobos hard line limit is 120 characters

BBasile commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4289,409 @@ unittest
+}
+
+///
+unittest
```
since the range primitive are marked as `pure nothrow @safe` you should also mark the unittest so, in order to verify that they really are

greenify commented:
@BBasile thanks for reviewing my code - I will try harder next time to stick to the coding style ;-)

Updates
- its now in "all allman"
- tests are in `pure nothrow @safe` - I replace `to!(dchar[])` with a couple of  postfix `d` s instead
- added unittests for binomial

quickfur commented:
IMO, `product` is a poor choice of name, especially since we already have `cartesianProduct` (does `product` mean a different kind of product?).  Maybe `cartesianPower` might be better? See also: https://issues.dlang.org/show_bug.cgi?id=7128

BBasile commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4289,478 @@ unittest
+            _max_states = length;
+            if (_length > 0 && _max_states > 0)
+            {
+                _state = new size_t[repeat];
```
Q: Why don't you use use the _.length_ array property here ?

greenify replied:
I do admit this is very confusing - the number of elements (`r.length`) is cached (in the documentation I read that it is _not_ guaranteed to be constant) and it is used quite often (in the `while` loop of `popFront`. 
Moreover `max_states` computes the all possible states as this makes computations a lot easier and provides the user with a nice `.length` property. 
To avoid this confusion I also renamed `lengthComputed` to `isNrStatesComputed`.

BBasile replied:
No I meant `_state.length = repeat;`, but maybe I miss something here (that why I've added the Q before).

greenify replied:
Oh never mind - better naming doesn't hurt ;-)
@BBasile you didn't miss anything - your comment is absolutely valid and changed it to this makes absolutely sense.
I am still not entirely used to D. Sorry & thanks!

greenify commented:
@quickfur 
> IMO, product is a poor choice of name, especially since we already have cartesianProduct (does 
product mean a different kind of product?)

I renamed it to `cartesianPower` and changed the documentation accordingly.
Moreover I also added the methods to the index header ;-)

andralex commented on std/algorithm/iteration.d:
```diff
@@ -11,11 +11,21 @@ $(T2 cache,
 $(T2 cacheBidirectional,
         As above, but also provides $(D back) and $(D popBack).)
+$(T2 cartesianPower,
+        Lazily computes the Cartesian product of $(D r) for a number of
```
I think the text is confusing, e.g. you want "Lazily computes the Cartesian product of `r` with itself..." The example should show a power of 3.

Also, I'd like more in terms of motivation. From the few people who need Cartesian products at all, a fraction would need the power of a range, and those could write 'cartesianProduct(r, r, r)`. True, that can't be done if the exponent is a variable. But how often does that happen?

andralex commented on std/algorithm/iteration.d:
```diff
@@ -11,11 +11,21 @@ $(T2 cache,
+$(T2 combinations,
+        Lazily computes all k-combinations of $(D r).
+        $(D"AB".combinations(2)) returns $(D["AB"]))
+$(T2 combinationsRepeat,
```
One is "combinations" and the other is "arrangements". No?

greenify replied:
Hmm "arrangement" is defined

> In general, an arrangement of objects is simply a grouping of them. The number of "arrangements" of n items is given either by a combination (order is ignored) or permutation (order is significant).

(from: http://mathworld.wolfram.com/Arrangement.html)

This is really about allowing repeats

> the number of ways to sample k elements from a set of n elements allowing for duplicates (i.e., with replacement) but disregarding different orderings (e.g. {2,1,2} = {1,2,2}).

See also:

https://en.wikipedia.org/wiki/Combination#Number_of_combinations_with_repetition
https://en.wikipedia.org/wiki/Multiset#Counting_multisets

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4299,484 @@ unittest
+    if (k > (n / 2))
+        k = n - k;
+    size_t result = 1;
+    for (size_t i=1, m=n; i<=k; i++, m--)
```
space around operators please - your own code does it elsewhere

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4299,484 @@ unittest
+*/
+auto cartesianPower(Range)(Range r, size_t repeat=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
```
brace on its own line

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4299,484 @@ unittest
+        private size_t _max_states, _pos;
+        private bool _isNrStatesComputed;
+
+        bool empty;
```
this is public and writable - not good

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4299,484 @@ unittest
+                ++el;
+                if (el < _nrElements)
+                    break;
+                else
```
redundant else

greenify replied:
why is this redundant? why can enter both states.
Maybe you we referring to the else in length, which was not needed ;-)

schuetzm replied:
Because the behaviour of the code doesn't change if you remove it. But IMO that's ok here, because it is more explicit.

greenify replied:
Hmm I do admit this is quite hard to see, but the `else` does change the behavior!
I added a lot of documentation, here's an excerpt!

```
Bitwise increment - starting from back
It works like adding 1 in primary school arithmetic.
If a block has reached the number of elements, we reset it to 0, and continue to increment, e.g. for n = 2:

[0, 0, 0] -> [0, 0, 1]
[0, 1, 1] -> [1, 0, 0]
```

schuetzm replied:
```D
for(;;) {
    if(cond) {
        blabla();
        break;
    } else {
        blubb();
    }
    foo();
}
```

is always equivalent to:

```D
for(;;) {
    if(cond) {
        blabla();
        break;
    }
    blubb();
    foo();
}
```

greenify replied:
ah now I finally got what you two where talking about :)

I prefer the more explicited way, but it's updated.

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4299,484 @@ unittest
+*/
+auto combinationsRepeat(Range)(Range r, size_t k=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
```
{

greenify commented:
> this is public and writable - not good

I also added a wrapper around `empty` for the existing permutation method.

Moreover I updated the code in regards to all your concerns (@andralex) - sorry about violating the code style.
I will try to be better next time ;-)

greenify commented:
> Also, I'd like more in terms of motivation. From the few people who need Cartesian products at all, a fraction would need the power of a range, and those could write 'cartesianProduct(r, r, r)`. True, that can't be done if the exponent is a variable. But how often does that happen?

At least for my use cases it is nearly always an variable.

I can argue that at least `product` is part of [Python's core itertools](https://docs.python.org/3/library/itertools.html#itertools.product), and for `itertools.product` one finds [__118,373__](https://github.com/search?q=itertools.product&type=Code&utf8=%E2%9C%93) matches on github. I guess the usage is even higher because - just `product` is valid too in python if correctly imported.

greenify commented:
Does anyone know why autotester is failing randomly with this message?

> make[1]: Leaving directory `/home/braddr/sandbox/at-client/pull-1978040-Linux_32_64/phobos'
timed out after 3600 seconds, step failed

@braddr ?

greenify commented:
Updates:

- I added a lot of comments to the sometimes hard to understand `popFront`'s method (all 3)
- I made `front` private and added a getter function
- I added many more unittests for `combinations` and `combinationsRepeat` -> it really turned out there was an edge case not covered in `combinationsRepeat` (when repeat was > number of elements)

greenify commented:
ping @andralex - what is your opinion on this addition? What is blocking this PR?

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4257,6 +4267,11 @@ struct Permutations(Range)
         next(2);
     }
+
+    @property bool empty() pure nothrow @safe
```
Not sure what this change is doing in this PR, but attributes for member functions of type templates are inferred, so explicit attributes are not necessary here (also `@nogc` is conspicuous by absence)

greenify replied:
> Not sure what this change is doing in this PR

This was done on @andralex [request](https://github.com/D-Programming-Language/phobos/pull/4026#discussion_r54889486) to make `empty` readonly - I thought if this applies for the methods below I thought it applies for `permutation` as well.

> also @nogc is conspicuous by absence

added.

JakobOvrum replied:
Please just use attribute inference. Explicit attributes are nice for documentation but we don't generally document range primitives.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+
+Params:
+    n = non negative integer
+    k = non negative integer, k <= n
```
While I think this is fine too, pre- and post-conditions are often put in their own sections, ala
```
Preconditions: `k <= n`
```
Note that since `n` and `k` are parameter names, all instances of these as standalone words in the documentation comment will be expanded as `$(D_PARAM …)`, which for dlang.org makes them italic. You can circumvent this by prepending the word with an underscore, like `_n`. You can compare the generated docs for what makes most sense.

greenify replied:
Thanks for the underscore hint - I added it to `n!/k!(n-k)`. 
I looked at a couple of examples in Phobos using preconditions and a typical value is `!empty` which isn't that helpful. I personally prefer to have extra conditions of variables at their explanation, so I left it. However I am open to more input on this ;-)
 
Btw ideally ddoc would parse the input contracts and display them.

JakobOvrum replied:
Yeah, I think either way is fine, whichever you like better.

It would be neat to have some reflection of input contracts in documentation, but input contracts are just arbitrarily complex code snippets, and even in simple cases there's syntactic baggage:
```d
assert(k <= n);
```
In this simple case the `assert(…);` would be noisy if reflected verbatim in generated documentation.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+*/
+
+
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
```
Does std.algorithm.iteration.binomial really make sense? Maybe std.math or std.numeric?

greenify replied:
Yeah I share your opinion (that's why I initially asked where to put this). Is putting it to `std.math` ok?

JakobOvrum replied:
I don't know which of std.math and std.numeric is most appropriate. @9il @ibuclaw?

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+Returns:
+    Forward range which yields the product items
+*/
+auto cartesianPower(Range)(Range r, size_t repeat=1)
```
Nitpick, space surrounding "operators" (i.e. `size_t repeat = 1`, although technically I don't think `=` is an operator here)

greenify replied:
thanks. corrected :)

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+For example $(D"AB".cartesianPower(2).array) returns $(D["AA", "AB", "BA", "BB"])
+
+Params:
+    r = RandomAccessRange or string as origin
```
I think "Random access range" reads better here. `RandomAccessRange` doesn't actually refer to any standard symbol. In at least one other case we use `$(REF_ALTTEXT Random access range, isRandomAccessRange, std, range, primitives)` for this kind of documentation.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+    import std.range: indexed;
+    static struct CartesianPower
+    {
+        private ElementType!Range[] _front;
```
For future reference, block attribute syntax ala `private:` can be used to DRY in these cases.

greenify replied:
thanks for the tip :)

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+
+        private bool _empty;
+
+        this(Range r, size_t repeat) pure nothrow @safe
```
Attributes for this constructor actually depend on the postblit operator for `Range`, so leave this to attribute inference

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+            return c;
+        }
+
+        @property ElementType!Range[] front() pure nothrow @nogc
```
Returns by value, so use attribute inference (again, postblit operator)

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+
+                el = 0;
+            }
+            _front = _r.indexed(_state).array;
```
Needs attribute inference

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+            return _max_states;
+        }
+
+        @property bool empty() pure nothrow @safe
```
Doesn't need attribute inference, but please either use it or specify all attributes explicitly, i.e. it's missing `@nogc`

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+
+/**
+Lazily computes all k-combinations of $(D r).
+Imagine this as the cartesianPower filtered for only strictly ordered items.
```
`$(LREF cartesianPower)` or "cartesian power"

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,625 @@ unittest
+{
+    import std.array: array;
+    import std.range: iota;
+    assert(iota(2).cartesianPower.array == [[0], [1]]);
```
It almost certainly won't make a measurable difference in trivial cases like this, but it's good practice in general to operate lazily when possible:
```d
assert(iota(2).cartesianPower.equal([[0], [1]]));
```
This is also pertinent for documented unit tests as they should espouse best practice.

---

The array literal causes another GC allocation which prevents `@nogc`, so include at least one unit test that tests `@nogc` by allocating the array statically:
```d
static immutable cheatsheet = [[0], [1]];
assert(iota(2).cartesianPower.equal(cheatsheet));
```

greenify commented:
> Please just use attribute inference. Explicit attributes are nice for documentation but we don't generally document range primitives.

Something learnt - thanks for the hint!

> but it's good practice in general to operate lazily when possible:

Done for all tests :)

> The array literal causes another GC allocation which prevents @nogc, so include at least one unit test that tests @nogc by allocating the array statically:

I am not quite sure whether `@nogc` will work out-of-the-box, as the state array is allocated depending on the length of k/repeat (`permutations` does the same).
However I removed all `.array` calls on `Indexed`, so we now every return element is of type `Indexed!(Range, size_t[])` and thus I had to use `equal!equal` for the tests - afaik `alias` doesn't work in unittests?

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,632 @@ unittest
+    static struct CartesianPower
+    {
+    private:
+        Indexed!(Range, size_t[]) _front;
```
I don't think we need to store this - just do `return _r.indexed(_state);` in `front`.

Since `_state` is bidirectional, which is propagated by `indexed`, it seems like `CartesianPower` should also be bidirectional?

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,632 @@ unittest
+        size_t _max_states, _pos;
+        bool _isNrStatesComputed;
+
+        bool _empty;
```
Can this be omitted by using `return _pos == _max_states;` in `empty`?

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,632 @@ unittest
+
+        size_t _repeat, _nrElements;
+        size_t _max_states, _pos;
+        bool _isNrStatesComputed;
```
This is set to true in the constructor (by calling `length`) and is true from there on - this doesn't seem necessary. Just move the calculation from `length` to the constructor

greenify replied:
> this doesn't seem necessary.

The main motivation was to avoid the redundant calculation when the struct is copied on `save` - is there a more elegant way to avoid that?

JakobOvrum replied:
I'm not sure what you mean. AFAICS, with the current setup the calculation is always done on `save`?

greenify replied:
> AFAICS, with the current setup the calculation is always done on save

Fair enough, we don't pass `max_states` nor _isNrStatesComputed. How about adding a 
private constructor that accepts an additional `max_states`?

JakobOvrum replied:
Sure, but you don't have to use a constructor. You can initialize the private fields directly in the body of `save`.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,627 @@ unittest
+            _max_states = length;
+            if (_nrElements > 0 && _max_states > 0)
+            {
+                _state.length = repeat;
```
Setting the length can do any of a number of things depending on the two inputs, so I think `_state = new size_t[](repeat);` would be clearer here

greenify commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,602 @@ unittest
+            if (empty) return;
+            _pos++;
+            // we might have bumped into the end state now
+            if (empty) return;
```
If `popFront` is called for the last state, we run below zero at the `while` loop below.

Alternatives:

- `adding `i > 0` to the loop
- `if (  _pos == _max_states - 1) { _pos++; return; }`  

Imho the end case happens seldom, so we want to avoid it if possible and for the other too algorithms we can do that.

greenify commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,602 @@ unittest
+        {
+            // check whether we have reached the end
+            if (empty) return;
+            _pos++;
```
saves an additional `if (empty)` [see](https://github.com/D-Programming-Language/phobos/pull/4026/files#r56037099)

greenify commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,602 @@ unittest
+        {
+            // check whether we have reached the end
+            if (empty) return;
+            _pos++;
```
saves an additional `if (empty)` [see](https://github.com/D-Programming-Language/phobos/pull/4026/files#r56037099)

greenify commented:
> I don't think we need to store this - just do return _r.indexed(_state); in front.

Updated.

> Can this be omitted by using return _pos == _max_states; in empty?

Initially I wanted to avoid that, because it's a comparison more. However the code looks much cleaner now :)

> Just move the calculation from length to the constructor

Removed `_isNrStatesComputed` and changed `save`  to copy the calculated props.

> Since _state is bidirectional, which is propagated by indexed, it seems like CartesianPower should also be bidirectional?

Yes it wouldn't be that difficult to add another state parameter and pos counter that traverses in the opposite direction. However will is this be useful for users, so that it justifies the overhead?

greenify commented:
> Since _state is bidirectional, which is propagated by indexed, it seems like CartesianPower should also be bidirectional?

@JakobOvrum I addded bidirectional support for CartesianPower - so that it's pro/cons can be better discussed. It should be easy to add the same to permutations, combinations and combinationsRepeat if wanted.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,633 @@ unittest
+        Range _r;
+        size_t[] _state, _stateBack;
+
+        size_t _repeat, _nrElements;
```
`_repeat` seems to never be used after the constructor (as it's identical to `_state.length` which is used instead)

---

`_nrElements` is `_r.length` - it's probably better to cache this as a local variable in the constructor and `pop*` functions respectively instead of as a field, i.e. as:
```d
void popFront()
{
    immutable nrElements = _r.length;
    /* use nrElements */
}
```
This is more in line with how other range algorithms do it. Chances are that `_r.length` is simply a struct field already, and when it isn't, it's probably cheap. We tend to err on the side of less struct state than aggressive caching.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,633 @@ unittest
+            if (_nrElements > 0)
+            {
+                _max_states = pow(_nrElements, _repeat);
+                if (_max_states > 0)
```
Since `_nrElements > 0` and `_repeat >= 1`, this seems like a safe assumption

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,633 @@ unittest
+
+        @property size_t length()
+        {
+            return _max_states;
```
The length must shrink as the range is consumed from the front or back; it doesn't seem to do that and it's not tested

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,633 @@ unittest
+}
+body
+{
+    import std.array: array;
```
This import is now unused AFAICS

greenify commented:
> The length must shrink as the range is consumed from the front or back; it doesn't seem to do that and it's not tested

Great point!

> it's probably better to cache this as a local variable in the constructor and pop* functions respectively instead of as a field

This feels a bit unnatural to me, but done ;-)

@JakobOvrum I covered also all other points and now I feel very ashamed that I made such a poor PR.

@9il If I read the documentation correctly `std.math` is only for elementary operations, so only `std.numeric` is left?

9il commented:
> @9il If I read the documentation correctly std.math is only for elementary operations, so only std.numeric is left?

Yes. In the same time, would you mind to place this functions to mir instead? `std.numeric` should will be deprecated in the future. We should think about new `sci.*` part of Phobos.

e10s commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
+in
+{
+    assert(n > 0, "binomial: n must be > 0.");
```
This `assert` is unnecessary because `n == 0` is also valid (as documented above).

greenify replied:
It now checks for `n >= 0`, there is a [generalization of the binomial coefficient for negative integers](https://en.wikipedia.org/wiki/Binomial_coefficient#Generalization_to_negative_integers) which is seldomly used - should it be added too?

greenify commented:
ping pong @andralex for general feedback on inclusion (see my [reasoning](https://github.com/D-Programming-Language/phobos/pull/4026#issuecomment-191815524) above).

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,592 @@ unittest
+        void popFront()
+        {
+            // check whether we have reached the end
+            if (empty) return;
```
Generally we use `assert(!empty);` in `popFront`, making it a logic error to call `popFront` on an empty range

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,592 @@ unittest
+        void popBack()
+        {
+            // check whether we have reached the end
+            if (empty) return;
```
Ditto

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4289,3 +4304,592 @@ unittest
+            {
+                _max_states = pow(r.length, repeat);
+                _state = new size_t[](repeat);
+                _stateBack = new size_t[](repeat);
```
When I suggested bidirectionality I was under the impression that it could be provided with the same underlying index array, but I see that's not the case. It should be reverted to just a forward range, and callers which need reverse iteration can simply retro the input:
```d
assert("AB"d.cartesianPower(2).equal!equal(["AA", "AB", "BA", "BB"]));
assert("AB"d.retro.cartesianPower(2).equal!equal(["BB", "BA", "AB", "AA"]));
```

JakobOvrum commented:
I'm no math guy but is there really no way to do these in constant space?

greenify commented:
> I'm no math guy but is there really no way to do these in constant space?

AFAIK having this state array (aka pools) is the most common way to do this - have a look at the implementation in Python:

Allocation: https://github.com/python/cpython/blob/520c7a75f1a600b3b3a57ced782dfd046ddff678/Modules/itertoolsmodule.c#L2041
Iteration: https://github.com/python/cpython/blob/520c7a75f1a600b3b3a57ced782dfd046ddff678/Modules/itertoolsmodule.c#L2161

(all other comments are addressed & updated)

greenify commented:
Could someone add the @andralex label? Thankls ;-)

9il commented:
Please add algorithms complexity overview

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+}
+body
+{
+    if (k < 0 || k > n)
```
k is unsigned

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
+in
+{
+    assert(n >= 0, "binomial: n must be non-negative");
```
n is unsigned

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+{
+    if (k < 0 || k > n)
+        return 0;
+    if (k == n)
```
no need to have this special case

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+*/
+
+public:
+size_t binomial(size_t n, size_t k) pure nothrow @safe @nogc
```
Should work with all unsigned types

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+{
+    assert(binomial(5, 1) == 5);
+    assert(binomial(5, 0) == 1);
+    assert(binomial(1, 2) == 0);
```
Test should contains big values

9il commented on std/numeric.d:
```diff
@@ -3088,3 +3088,56 @@ unittest
+        k = n - k;
+    size_t result = 1;
+    for (size_t i = 1, m = n; i <= k; i++, m--)
+        result = result * m / i;
```
any proof that this recursion is correct? Why not to use the recursive formula from wikipedia without multiplication and division?

9il replied:
Oh, recursive has bad complexity.

9il replied:
From wikipedia:
```C++
#include <limits.h>
unsigned binomial(unsigned n, unsigned k) {
  unsigned c = 1, i;
  if (k > n-k) k = n-k;  /* take advantage of symmetry */
  for (i = 1; i <= k; i++, n--) {
    if (c/i > UINT_MAX/n) return 0;  /* return 0 on overflow */
    c = c/i * n + c%i * n / i;  /* split c*n/i into (c/i*i + c%i)*n/i */
  }
  return c;
}
```

9il commented:
@greenify We need an experiments with this functions before including them to Phobos.
Please include all this functions to your Mir PR, update module name with `mir.combinatorics` (or another name, not `stat`), fix issues with `binomial` I have listed here, and close this PR.

9il commented:
In addition, it is possible to compute FP `binomial` value with minimal rounding error using precise summation algorithm.

greenify commented:
> Please include all this functions to your Mir PR, update module name with mir.combinatorics (or another name, not stat), fix issues with binomial I have listed here, and close this PR.

[Done](https://github.com/DlangScience/mir/pull/29) - even though I would prefer to have these iteration  algorithms in `std.algorithm.iteration` eventually ;-)

wilzbach commented:
Update: 15 days and a lot of [review & help](https://github.com/DlangScience/mir/pull/29) (thanks so much @9il) later - an initial version of `combinatorics` is now part of mir. It now supports `@nogc` via the Allocator API.

See [combinatorics](https://github.com/DlangScience/mir/blob/master/source/mir/combinatorics/package.d) at [mir](https://github.com/DlangScience/mir)
